File: imessage_cli/watcher.py
"""Database watcher for real-time message updates."""

import os
import sqlite3
import threading
import time
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path
from typing import Callable, List, Optional

from .database import get_db_path, get_connection, apple_time_to_datetime, extract_text_from_attributed_body


@dataclass
class Message:
    """Represents an iMessage."""
    message_id: int
    text: str
    date: Optional[datetime]
    is_from_me: bool
    is_read: bool
    sender: str
    chat_id: int
    chat_identifier: str
    chat_name: str


@dataclass
class Conversation:
    """Represents a conversation/chat."""
    chat_id: int
    chat_identifier: str
    display_name: str
    service: str
    last_message_date: Optional[datetime]
    last_message_text: str
    unread_count: int
    participants: List[str]


class MessageWatcher:
    """Watches the iMessage database for new messages."""
    
    def __init__(self, poll_interval: float = 1.0):
        """Initialize the watcher.
        
        Args:
            poll_interval: How often to check for new messages (seconds)
        """
        self.poll_interval = poll_interval
        self._running = False
        self._thread: Optional[threading.Thread] = None
        self._last_message_id = 0
        self._last_mtime = 0.0
        self._callbacks: List[Callable[[List[Message]], None]] = []
        self._conversation_callbacks: List[Callable[[List[Conversation]], None]] = []
        self._error_callbacks: List[Callable[[Exception], None]] = []
        
    def on_new_messages(self, callback: Callable[[List[Message]], None]):
        """Register a callback for new messages."""
        self._callbacks.append(callback)
        
    def on_conversations_updated(self, callback: Callable[[List[Conversation]], None]):
        """Register a callback for conversation updates."""
        self._conversation_callbacks.append(callback)
        
    def on_error(self, callback: Callable[[Exception], None]):
        """Register a callback for errors."""
        self._error_callbacks.append(callback)
    
    def _get_last_message_id(self) -> int:
        """Get the ID of the most recent message."""
        try:
            conn = get_connection()
            cursor = conn.cursor()
            cursor.execute("SELECT MAX(ROWID) FROM message")
            result = cursor.fetchone()
            conn.close()
            return result[0] or 0
        except Exception:
            return 0
    
    def _get_db_mtime(self) -> float:
        """Get the modification time of the database file."""
        try:
            db_path = get_db_path()
            return os.path.getmtime(db_path)
        except Exception:
            return 0.0
    
    def get_conversations(self, limit: int = 50) -> List[Conversation]:
        """Get list of conversations with metadata."""
        conn = get_connection()
        cursor = conn.cursor()
        
        query = """
        SELECT 
            c.ROWID as chat_id,
            c.chat_identifier,
            c.display_name,
            c.service_name,
            MAX(m.date) as last_message_date,
            (SELECT text FROM message m2 
             JOIN chat_message_join cmj2 ON m2.ROWID = cmj2.message_id 
             WHERE cmj2.chat_id = c.ROWID 
             ORDER BY m2.date DESC LIMIT 1) as last_message_text,
            SUM(CASE WHEN m.is_read = 0 AND m.is_from_me = 0 THEN 1 ELSE 0 END) as unread_count,
            GROUP_CONCAT(DISTINCT h.id) as participants
        FROM chat c
        LEFT JOIN chat_message_join cmj ON c.ROWID = cmj.chat_id
        LEFT JOIN message m ON cmj.message_id = m.ROWID
        LEFT JOIN chat_handle_join chj ON c.ROWID = chj.chat_id
        LEFT JOIN handle h ON chj.handle_id = h.ROWID
        GROUP BY c.ROWID
        ORDER BY last_message_date DESC
        LIMIT ?
        """
        
        cursor.execute(query, (limit,))
        rows = cursor.fetchall()
        conn.close()
        
        conversations = []
        for row in rows:
            last_date = apple_time_to_datetime(row[4])
            conversations.append(Conversation(
                chat_id=row[0],
                chat_identifier=row[1],
                display_name=row[2] or row[1] or "Unknown",
                service=row[3] or "iMessage",
                last_message_date=last_date,
                last_message_text=row[5] or "",
                unread_count=row[6] or 0,
                participants=row[7].split(',') if row[7] else []
            ))
        
        return conversations
    
    def get_messages(self, chat_id: int, limit: int = 100) -> List[Message]:
        """Get messages for a specific chat."""
        conn = get_connection()
        cursor = conn.cursor()
        
        query = """
        SELECT 
            m.ROWID as message_id,
            m.text,
            m.attributedBody,
            m.date,
            m.is_from_me,
            m.is_read,
            h.id as sender_id,
            c.ROWID as chat_id,
            c.chat_identifier,
            c.display_name
        FROM message m
        LEFT JOIN chat_message_join cmj ON m.ROWID = cmj.message_id
        LEFT JOIN chat c ON cmj.chat_id = c.ROWID
        LEFT JOIN handle h ON m.handle_id = h.ROWID
        WHERE c.ROWID = ?
        ORDER BY m.date DESC
        LIMIT ?
        """
        
        cursor.execute(query, (chat_id, limit))
        rows = cursor.fetchall()
        conn.close()
        
        messages = []
        for row in rows:
            msg_date = apple_time_to_datetime(row[3])
            
            # Try to get text from the text column first, then fall back to attributedBody
            text = row[1]
            if not text and row[2]:
                text = extract_text_from_attributed_body(row[2])
            if not text:
                text = "[Attachment]"
            
            messages.append(Message(
                message_id=row[0],
                text=text,
                date=msg_date,
                is_from_me=bool(row[4]),
                is_read=bool(row[5]),
                sender='Me' if row[4] else (row[6] or 'Unknown'),
                chat_id=row[7],
                chat_identifier=row[8],
                chat_name=row[9] or row[8] or "Unknown"
            ))
        
        # Return in chronological order
        messages.reverse()
        return messages
    
    def get_new_messages(self, since_id: int) -> List[Message]:
        """Get messages newer than the given ID."""
        conn = get_connection()
        cursor = conn.cursor()
        
        query = """
        SELECT 
            m.ROWID as message_id,
            m.text,
            m.attributedBody,
            m.date,
            m.is_from_me,
            m.is_read,
            h.id as sender_id,
            c.ROWID as chat_id,
            c.chat_identifier,
            c.display_name
        FROM message m
        LEFT JOIN chat_message_join cmj ON m.ROWID = cmj.message_id
        LEFT JOIN chat c ON cmj.chat_id = c.ROWID
        LEFT JOIN handle h ON m.handle_id = h.ROWID
        WHERE m.ROWID > ?
        ORDER BY m.date ASC
        """
        
        cursor.execute(query, (since_id,))
        rows = cursor.fetchall()
        conn.close()
        
        messages = []
        for row in rows:
            msg_date = apple_time_to_datetime(row[3])
            
            # Try to get text from the text column first, then fall back to attributedBody
            text = row[1]
            if not text and row[2]:
                text = extract_text_from_attributed_body(row[2])
            if not text:
                text = "[Attachment]"
            
            messages.append(Message(
                message_id=row[0],
                text=text,
                date=msg_date,
                is_from_me=bool(row[4]),
                is_read=bool(row[5]),
                sender='Me' if row[4] else (row[6] or 'Unknown'),
                chat_id=row[7],
                chat_identifier=row[8],
                chat_name=row[9] or row[8] or "Unknown"
            ))
        
        return messages
    
    def _poll_loop(self):
        """Main polling loop."""
        while self._running:
            try:
                # Check if database has been modified
                current_mtime = self._get_db_mtime()
                
                if current_mtime > self._last_mtime:
                    self._last_mtime = current_mtime
                    
                    # Check for new messages
                    current_max_id = self._get_last_message_id()
                    
                    if current_max_id > self._last_message_id:
                        new_messages = self.get_new_messages(self._last_message_id)
                        self._last_message_id = current_max_id
                        
                        if new_messages:
                            for callback in self._callbacks:
                                try:
                                    callback(new_messages)
                                except Exception as e:
                                    self._notify_error(e)
                    
                    # Update conversations
                    conversations = self.get_conversations()
                    for callback in self._conversation_callbacks:
                        try:
                            callback(conversations)
                        except Exception as e:
                            self._notify_error(e)
                            
            except Exception as e:
                self._notify_error(e)
            
            time.sleep(self.poll_interval)
    
    def _notify_error(self, error: Exception):
        """Notify error callbacks."""
        for callback in self._error_callbacks:
            try:
                callback(error)
            except Exception:
                pass
    
    def start(self):
        """Start watching for new messages."""
        if self._running:
            return
        
        # Initialize state
        self._last_message_id = self._get_last_message_id()
        self._last_mtime = self._get_db_mtime()
        self._running = True
        
        self._thread = threading.Thread(target=self._poll_loop, daemon=True)
        self._thread.start()
    
    def stop(self):
        """Stop watching for messages."""
        self._running = False
        if self._thread:
            self._thread.join(timeout=2.0)
            self._thread = None
    
    def __enter__(self):
        self.start()
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        self.stop()
        return False

File: go/internal/watcher/watcher.go
// Package watcher provides real-time message watching functionality.
package watcher

import (
	"database/sql"
	"log"
	"os"
	"sync"
	"sync/atomic"
	"time"

	"github.com/danewalton/imessage-cli/internal/database"
)

// Watcher constants
const (
	DefaultPollInterval      = 500 * time.Millisecond
	DefaultConversationLimit = 50
)

// Message represents an iMessage for the watcher.
type Message struct {
	MessageID      int64
	Text           string
	Date           *time.Time
	IsFromMe       bool
	IsRead         bool
	Sender         string
	ChatID         int64
	ChatIdentifier string
	ChatName       string
}

// Conversation represents a conversation for the watcher.
type Conversation struct {
	ChatID          int64
	ChatIdentifier  string
	DisplayName     string
	Service         string
	LastMessageDate *time.Time
	LastMessageText string
	UnreadCount     int
	Participants    []string
}

// MessageCallback is called when new messages arrive.
type MessageCallback func([]Message)

// ConversationCallback is called when conversations are updated.
type ConversationCallback func([]Conversation)

// ErrorCallback is called when an error occurs.
type ErrorCallback func(error)

// MessageWatcher watches the iMessage database for new messages.
type MessageWatcher struct {
	pollInterval          time.Duration
	running               bool
	lastMessageID         atomic.Int64
	lastMtime             atomic.Int64
	messageCallbacks      []MessageCallback
	conversationCallbacks []ConversationCallback
	errorCallbacks        []ErrorCallback
	mu                    sync.RWMutex
	stopCh                chan struct{}
	wg                    sync.WaitGroup
	// logger for debugging callback issues
	logger *log.Logger
}

// NewMessageWatcher creates a new MessageWatcher.
func NewMessageWatcher(pollInterval time.Duration) *MessageWatcher {
	return &MessageWatcher{
		pollInterval: pollInterval,
		stopCh:       make(chan struct{}),
	}
}

// OnNewMessages registers a callback for new messages.
func (w *MessageWatcher) OnNewMessages(callback MessageCallback) {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.messageCallbacks = append(w.messageCallbacks, callback)
}

// OnConversationsUpdated registers a callback for conversation updates.
func (w *MessageWatcher) OnConversationsUpdated(callback ConversationCallback) {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.conversationCallbacks = append(w.conversationCallbacks, callback)
}

// OnError registers a callback for errors.
func (w *MessageWatcher) OnError(callback ErrorCallback) {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.errorCallbacks = append(w.errorCallbacks, callback)
}

func (w *MessageWatcher) getLastMessageID() int64 {
	db, err := database.GetConnection()
	if err != nil {
		return 0
	}
	defer db.Close()

	var maxID sql.NullInt64
	err = db.QueryRow("SELECT MAX(ROWID) FROM message").Scan(&maxID)
	if err != nil || !maxID.Valid {
		return 0
	}
	return maxID.Int64
}

func (w *MessageWatcher) getDBMtime() int64 {
	dbPath := database.GetDBPath()
	info, err := os.Stat(dbPath)
	if err != nil {
		return 0
	}
	return info.ModTime().UnixNano()
}

// GetConversations returns a list of conversations.
func (w *MessageWatcher) GetConversations(limit int) []Conversation {
	convs, err := database.GetConversations(limit)
	if err != nil {
		return nil
	}

	var result []Conversation
	for _, c := range convs {
		result = append(result, Conversation{
			ChatID:          c.ChatID,
			ChatIdentifier:  c.ChatIdentifier,
			DisplayName:     c.DisplayName,
			Service:         c.Service,
			LastMessageDate: c.LastMessageDate,
			LastMessageText: c.LastMessageText,
			UnreadCount:     c.UnreadCount,
			Participants:    c.Participants,
		})
	}
	return result
}

// GetMessages returns messages for a specific chat.
func (w *MessageWatcher) GetMessages(chatID int64, limit int) []Message {
	msgs, err := database.GetMessages(chatID, "", limit)
	if err != nil {
		return nil
	}

	var result []Message
	for _, m := range msgs {
		result = append(result, Message{
			MessageID:      m.MessageID,
			Text:           m.Text,
			Date:           m.Date,
			IsFromMe:       m.IsFromMe,
			IsRead:         m.IsRead,
			Sender:         m.Sender,
			ChatID:         m.ChatID,
			ChatIdentifier: m.ChatIdent,
			ChatName:       m.ChatName,
		})
	}
	return result
}

// GetNewMessages returns messages newer than the given ID.
func (w *MessageWatcher) GetNewMessages(sinceID int64) []Message {
	db, err := database.GetConnection()
	if err != nil {
		w.notifyError(err)
		return nil
	}
	defer db.Close()

	query := `
		SELECT 
			m.ROWID as message_id,
			m.text,
			m.attributedBody,
			m.date,
			m.is_from_me,
			m.is_read,
			h.id as sender_id,
			c.ROWID as chat_id,
			c.chat_identifier,
			c.display_name
		FROM message m
		LEFT JOIN chat_message_join cmj ON m.ROWID = cmj.message_id
		LEFT JOIN chat c ON cmj.chat_id = c.ROWID
		LEFT JOIN handle h ON m.handle_id = h.ROWID
		WHERE m.ROWID > ?
		ORDER BY m.date ASC
	`

	rows, err := db.Query(query, sinceID)
	if err != nil {
		return nil
	}
	defer rows.Close()

	var messages []Message
	for rows.Next() {
		var m Message
		var text, senderID, chatIdent, chatName sql.NullString
		var attributedBody []byte
		var date sql.NullInt64
		var isFromMe, isRead int

		err := rows.Scan(&m.MessageID, &text, &attributedBody, &date, &isFromMe, &isRead, &senderID, &m.ChatID, &chatIdent, &chatName)
		if err != nil {
			continue
		}

		m.IsFromMe = isFromMe == 1
		m.IsRead = isRead == 1
		m.ChatIdentifier = chatIdent.String
		m.ChatName = chatName.String

		if date.Valid {
			m.Date = database.AppleTimeToTime(date.Int64)
		}

		m.Text = text.String
		if m.Text == "" && len(attributedBody) > 0 {
			m.Text = database.ExtractTextFromAttributedBody(attributedBody)
		}
		if m.Text == "" {
			m.Text = "[Attachment]"
		}

		m.Sender = database.ResolveSender(m.IsFromMe, senderID.String)

		if m.ChatName == "" {
			m.ChatName = database.GetContactName(m.ChatIdentifier)
		}

		messages = append(messages, m)
	}

	return messages
}

func (w *MessageWatcher) pollLoop() {
	defer w.wg.Done()

	ticker := time.NewTicker(w.pollInterval)
	defer ticker.Stop()

	for {
		select {
		case <-w.stopCh:
			return
		case <-ticker.C:
			w.poll()
		}
	}
}

func (w *MessageWatcher) poll() {
	// Check if database has been modified
	currentMtime := w.getDBMtime()
	lastMtime := w.lastMtime.Load()

	if currentMtime > lastMtime {
		w.lastMtime.Store(currentMtime)

		// Check for new messages
		currentMaxID := w.getLastMessageID()
		lastID := w.lastMessageID.Load()

		if currentMaxID > lastID {
			newMessages := w.GetNewMessages(lastID)
			w.lastMessageID.Store(currentMaxID)

			if len(newMessages) > 0 {
				w.mu.RLock()
				callbacks := make([]MessageCallback, len(w.messageCallbacks))
				copy(callbacks, w.messageCallbacks)
				w.mu.RUnlock()

				for _, cb := range callbacks {
					go func(callback MessageCallback, msgs []Message) {
						defer func() {
							if r := recover(); r != nil {
								if w.logger != nil {
									w.logger.Printf("panic in message callback: %v", r)
								}
							}
						}()
						callback(msgs)
					}(cb, newMessages)
				}
			}
		}

		// Update conversations
		conversations := w.GetConversations(DefaultConversationLimit)
		w.mu.RLock()
		callbacks := make([]ConversationCallback, len(w.conversationCallbacks))
		copy(callbacks, w.conversationCallbacks)
		w.mu.RUnlock()

		for _, cb := range callbacks {
			go func(callback ConversationCallback, convs []Conversation) {
				defer func() {
					if r := recover(); r != nil {
						if w.logger != nil {
							w.logger.Printf("panic in conversation callback: %v", r)
						}
					}
				}()
				callback(convs)
			}(cb, conversations)
		}
	}
}

func (w *MessageWatcher) notifyError(err error) {
	w.mu.RLock()
	defer w.mu.RUnlock()
	for _, cb := range w.errorCallbacks {
		go cb(err)
	}
}

// Start begins watching for new messages.
func (w *MessageWatcher) Start() {
	w.mu.Lock()
	if w.running {
		w.mu.Unlock()
		return
	}

	// Mark running and create stop channel immediately to avoid blocking
	w.running = true
	w.stopCh = make(chan struct{})
	w.mu.Unlock()

	// Start poll loop in a goroutine; perform initial DB checks there to avoid blocking caller
	w.wg.Add(1)
	go func() {
		// Initialize last IDs / mtime inside goroutine using atomic operations
		w.lastMessageID.Store(w.getLastMessageID())
		w.lastMtime.Store(w.getDBMtime())

		w.pollLoop()
	}()
}

// Stop stops watching for messages.
func (w *MessageWatcher) Stop() {
	w.mu.Lock()
	if !w.running {
		w.mu.Unlock()
		return
	}
	w.running = false
	close(w.stopCh)
	w.mu.Unlock()

	w.wg.Wait()
}
